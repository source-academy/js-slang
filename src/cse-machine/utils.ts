import * as es from 'estree'
import { uniqueId } from 'lodash'

import { Context } from '..'
import * as errors from '../errors/errors'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import Closure from '../interpreter/closure'
import { Environment, Frame, Value } from '../types'
import * as ast from '../utils/astCreator'
import * as instr from './instrCreator'
import { Agenda } from './interpreter'
import { AgendaItem, AppInstr, AssmtInstr, Instr, InstrType } from './types'

/**
 * Stack is implemented for agenda and stash registers.
 */
interface IStack<T> {
  push(...items: T[]): void
  pop(): T | undefined
  peek(): T | undefined
  size(): number
  isEmpty(): boolean
  getStack(): T[]
}

export class Stack<T> implements IStack<T> {
  // Bottom of the array is at index 0
  private storage: T[] = []

  public constructor() {}

  public push(...items: T[]): void {
    for (const item of items) {
      this.storage.push(item)
    }
  }

  public pop(): T | undefined {
    return this.storage.pop()
  }

  public peek(): T | undefined {
    if (this.isEmpty()) {
      return undefined
    }
    return this.storage[this.size() - 1]
  }

  public size(): number {
    return this.storage.length
  }

  public isEmpty(): boolean {
    return this.size() == 0
  }

  public getStack(): T[] {
    // return a copy of the stack's contents
    return [...this.storage]
  }

  public some(predicate: (value: T) => boolean): boolean {
    return this.storage.some(predicate)
  }
}

/**
 * Typeguard for Instr to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an instruction and false otherwise.
 */
export const isInstr = (command: AgendaItem): command is Instr => {
  return (command as Instr).instrType !== undefined
}

/**
 * Typeguard for esNode to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an esNode and false if it is an instruction.
 */
export const isNode = (command: AgendaItem): command is es.Node => {
  return (command as es.Node).type !== undefined
}

/**
 * Typeguard for esIdentifier. To verify if an esNode is an esIdentifier.
 *
 * @param node an esNode
 * @returns true if node is an esIdentifier, false otherwise.
 */
export const isIdentifier = (node: es.Node): node is es.Identifier => {
  return (node as es.Identifier).name !== undefined
}

/**
 * Typeguard for esReturnStatement. To verify if an esNode is an esReturnStatement.
 *
 * @param node an esNode
 * @returns true if node is an esReturnStatement, false otherwise.
 */
export const isReturnStatement = (node: es.Node): node is es.ReturnStatement => {
  return (node as es.ReturnStatement).type == 'ReturnStatement'
}

/**
 * Typeguard for esIfStatement. To verify if an esNode is an esIfStatement.
 *
 * @param node an esNode
 * @returns true if node is an esIfStatement, false otherwise.
 */
export const isIfStatement = (node: es.Node): node is es.IfStatement => {
  return (node as es.IfStatement).type == 'IfStatement'
}

/**
 * Typeguard for esBlockStatement. To verify if an esNode is a block statement.
 *
 * @param node an esNode
 * @returns true if node is an esBlockStatement, false otherwise.
 */
export const isBlockStatement = (node: es.Node): node is es.BlockStatement => {
  return (node as es.BlockStatement).type == 'BlockStatement'
}

/**
 * Typeguard for esRestElement. To verify if an esNode is a block statement.
 *
 * @param node an esNode
 * @returns true if node is an esRestElement, false otherwise.
 */
export const isRestElement = (node: es.Node): node is es.RestElement => {
  return (node as es.RestElement).type == 'RestElement'
}

/**
 * Typeguard for AssmtInstr. To verify if an instruction is an assignment instruction.
 *
 * @param instr an instruction
 * @returns true if instr is an AssmtInstr, false otherwise.
 */
export const isAssmtInstr = (instr: Instr): instr is AssmtInstr => {
  return instr.instrType === InstrType.ASSIGNMENT
}

/**
 * A helper function for handling sequences of statements.
 * Statements must be pushed in reverse order, and each statement is separated by a pop
 * instruction so that only the result of the last statement remains on stash.
 * Value producing statements have an extra pop instruction.
 *
 * @param seq Array of statements.
 * @returns Array of commands to be pushed into agenda.
 */
export const handleSequence = (seq: es.Statement[]): AgendaItem[] => {
  const result: AgendaItem[] = []
  let valueProduced = false
  for (const command of seq) {
    if (!isImportDeclaration(command)) {
      if (valueProducing(command)) {
        // Value producing statements have an extra pop instruction
        if (valueProduced) {
          result.push(instr.popInstr(command))
        } else {
          valueProduced = true
        }
      }
      result.push(command)
    }
  }
  // Push statements in reverse order
  return result.reverse()
}

/**
 * This function is used for ConditionalExpressions and IfStatements, to create the sequence
 * of agenda items to be added.
 */
export const reduceConditional = (
  node: es.IfStatement | es.ConditionalExpression
): AgendaItem[] => {
  return [instr.branchInstr(node.consequent, node.alternate, node), node.test]
}

/**
 * To determine if an agenda item is value producing. JavaScript distinguishes value producing
 * statements and non-value producing statements.
 * Refer to https://sourceacademy.nus.edu.sg/sicpjs/4.1.2 exercise 4.8.
 *
 * @param command Agenda item to determine if it is value producing.
 * @returns true if it is value producing, false otherwise.
 */
export const valueProducing = (command: es.Node): boolean => {
  const type = command.type
  return (
    type !== 'VariableDeclaration' &&
    type !== 'FunctionDeclaration' &&
    type !== 'ContinueStatement' &&
    type !== 'BreakStatement' &&
    type !== 'DebuggerStatement' &&
    (type !== 'BlockStatement' || command.body.some(valueProducing))
  )
}

/**
 * To determine if an agenda item changes the environment.
 * There is a change in the environment when
 *  1. pushEnvironment() is called when creating a new frame, if there are variable declarations.
 *     Called in Program, BlockStatement, and Application instructions.
 *  2. there is an assignment.
 *     Called in Assignment and Array Assignment instructions.
 *
 * @param command Agenda item to check against.
 * @returns true if it changes the environment, false otherwise.
 */
export const envChanging = (command: AgendaItem): boolean => {
  if (isNode(command)) {
    const type = command.type
    return type === 'Program' || (type === 'BlockStatement' && hasDeclarations(command))
  } else {
    const type = command.instrType
    return (
      type === InstrType.ASSIGNMENT ||
      type === InstrType.ARRAY_ASSIGNMENT ||
      (type === InstrType.APPLICATION && (command as AppInstr).numOfArgs > 0)
    )
  }
}

/**
 * To determine if the function is simple.
 * Simple functions contain a single return statement.
 *
 * @param node The function to check against.
 * @returns true if the function is simple, false otherwise.
 */
export const isSimpleFunction = (node: es.Function) => {
  if (node.body.type !== 'BlockStatement') {
    return true
  } else {
    const block = node.body
    return block.body.length === 1 && block.body[0].type === 'ReturnStatement'
  }
}

/**
 * Environments
 */

export const currentEnvironment = (context: Context) => context.runtime.environments[0]

export const createEnvironment = (
  closure: Closure,
  args: Value[],
  callExpression: es.CallExpression
): Environment => {
  const environment: Environment = {
    name: isIdentifier(callExpression.callee) ? callExpression.callee.name : closure.functionName,
    tail: closure.environment,
    head: {},
    id: uniqueId(),
    callExpression: {
      ...callExpression,
      arguments: args.map(ast.primitive)
    }
  }
  closure.node.params.forEach((param, index) => {
    if (isRestElement(param)) {
      environment.head[(param.argument as es.Identifier).name] = args.slice(index)
    } else {
      environment.head[(param as es.Identifier).name] = args[index]
    }
  })
  return environment
}

export const popEnvironment = (context: Context) => context.runtime.environments.shift()

export const pushEnvironment = (context: Context, environment: Environment) => {
  context.runtime.environments.unshift(environment)
  context.runtime.environmentTree.insert(environment)
}

export const createBlockEnvironment = (
  context: Context,
  name = 'blockEnvironment',
  head: Frame = {}
): Environment => {
  return {
    name,
    tail: currentEnvironment(context),
    head,
    id: uniqueId()
  }
}

/**
 * Variables
 */

const DECLARED_BUT_NOT_YET_ASSIGNED = Symbol('Used to implement block scope')

export function declareIdentifier(
  context: Context,
  name: string,
  node: es.Node,
  environment: Environment
) {
  if (environment.head.hasOwnProperty(name)) {
    const descriptors = Object.getOwnPropertyDescriptors(environment.head)

    return handleRuntimeError(
      context,
      new errors.VariableRedeclaration(node, name, descriptors[name].writable)
    )
  }
  environment.head[name] = DECLARED_BUT_NOT_YET_ASSIGNED
  return environment
}

function declareVariables(
  context: Context,
  node: es.VariableDeclaration,
  environment: Environment
) {
  for (const declaration of node.declarations) {
    declareIdentifier(context, (declaration.id as es.Identifier).name, node, environment)
  }
}

export function declareFunctionsAndVariables(
  context: Context,
  node: es.BlockStatement,
  environment: Environment
) {
  for (const statement of node.body) {
    switch (statement.type) {
      case 'VariableDeclaration':
        declareVariables(context, statement, environment)
        break
      case 'FunctionDeclaration':
        declareIdentifier(context, (statement.id as es.Identifier).name, statement, environment)
        break
    }
  }
}

export function hasDeclarations(node: es.BlockStatement): boolean {
  for (const statement of node.body) {
    if (statement.type === 'VariableDeclaration' || statement.type === 'FunctionDeclaration') {
      return true
    }
  }
  return false
}

export function hasImportDeclarations(node: es.BlockStatement): boolean {
  for (const statement of (node as unknown as es.Program).body) {
    if (statement.type === 'ImportDeclaration') {
      return true
    }
  }
  return false
}

function isImportDeclaration(node: es.Node): boolean {
  return node.type === 'ImportDeclaration'
}

export function defineVariable(
  context: Context,
  name: string,
  value: Value,
  constant = false,
  node: es.VariableDeclaration | es.ImportDeclaration
) {
  const environment = currentEnvironment(context)

  if (environment.head[name] !== DECLARED_BUT_NOT_YET_ASSIGNED) {
    return handleRuntimeError(context, new errors.VariableRedeclaration(node, name, !constant))
  }

  Object.defineProperty(environment.head, name, {
    value,
    writable: !constant,
    enumerable: true
  })

  return environment
}

export const getVariable = (context: Context, name: string, node: es.Identifier) => {
  let environment: Environment | null = currentEnvironment(context)
  while (environment) {
    if (environment.head.hasOwnProperty(name)) {
      if (environment.head[name] === DECLARED_BUT_NOT_YET_ASSIGNED) {
        return handleRuntimeError(context, new errors.UnassignedVariable(name, node))
      } else {
        return environment.head[name]
      }
    } else {
      environment = environment.tail
    }
  }
  return handleRuntimeError(context, new errors.UndefinedVariable(name, node))
}

export const setVariable = (
  context: Context,
  name: string,
  value: any,
  node: es.AssignmentExpression
) => {
  let environment: Environment | null = currentEnvironment(context)
  while (environment) {
    if (environment.head.hasOwnProperty(name)) {
      if (environment.head[name] === DECLARED_BUT_NOT_YET_ASSIGNED) {
        break
      }
      const descriptors = Object.getOwnPropertyDescriptors(environment.head)
      if (descriptors[name].writable) {
        environment.head[name] = value
        return undefined
      }
      return handleRuntimeError(context, new errors.ConstAssignment(node, name))
    } else {
      environment = environment.tail
    }
  }
  return handleRuntimeError(context, new errors.UndefinedVariable(name, node))
}

export const handleRuntimeError = (context: Context, error: RuntimeSourceError) => {
  context.errors.push(error)
  throw error
}

export const checkNumberOfArguments = (
  context: Context,
  callee: Closure | Value,
  args: Value[],
  exp: es.CallExpression
) => {
  if (callee instanceof Closure) {
    // User-defined or Pre-defined functions
    const params = callee.node.params
    const hasVarArgs = params[params.length - 1]?.type === 'RestElement'
    if (hasVarArgs ? params.length - 1 > args.length : params.length !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(
          exp,
          hasVarArgs ? params.length - 1 : params.length,
          args.length,
          hasVarArgs
        )
      )
    }
  } else {
    // Pre-built functions
    const hasVarArgs = callee.minArgsNeeded != undefined
    if (hasVarArgs ? callee.minArgsNeeded > args.length : callee.length !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(
          exp,
          hasVarArgs ? callee.minArgsNeeded : callee.length,
          args.length,
          hasVarArgs
        )
      )
    }
  }
  return undefined
}

/**
 * This function can be used to check for a stack overflow.
 * The current limit is set to be an agenda size of 1.0 x 10^5, if the agenda
 * flows beyond this limit an error is thrown.
 * This corresponds to about 10mb of space according to tests ran.
 */
export const checkStackOverFlow = (context: Context, agenda: Agenda) => {
  if (agenda.size() > 100000) {
    const stacks: es.CallExpression[] = []
    let counter = 0
    for (
      let i = 0;
      counter < errors.MaximumStackLimitExceeded.MAX_CALLS_TO_SHOW &&
      i < context.runtime.environments.length;
      i++
    ) {
      if (context.runtime.environments[i].callExpression) {
        stacks.unshift(context.runtime.environments[i].callExpression!)
        counter++
      }
    }
    handleRuntimeError(
      context,
      new errors.MaximumStackLimitExceeded(context.runtime.nodes[0], stacks)
    )
  }
}

/**
 * Checks whether a function body returns in every possible branch.
 * Returns true if every branch has a return statement, else returns false.
 * @param body The function body to be checked
 */
export const hasReturnStatement = (body: es.Statement): boolean => {
  if (!isBlockStatement(body)) return isReturnStatement(body)
  for (const statement of body.body) {
    if (isReturnStatement(statement)) {
      return true
    }
    if (isIfStatement(statement)) {
      const consequent = hasReturnStatement(statement.consequent)
      if (!consequent) {
        return false
      }
      if (statement.alternate) {
        return hasReturnStatement(statement.alternate)
      }
    }
  }
  return false
}

export const hasBreakStatement = (block: es.BlockStatement): boolean => {
  let hasBreak = false
  for (const statement of block.body) {
    if (statement.type === 'BreakStatement') {
      hasBreak = true
    } else if (statement.type === 'IfStatement') {
      // Parser enforces that if/else have braces (block statement)
      hasBreak = hasBreak || hasBreakStatement(statement.consequent as es.BlockStatement)
      if (statement.alternate) {
        hasBreak = hasBreak || hasBreakStatement(statement.alternate as es.BlockStatement)
      }
    }
  }
  return hasBreak
}

export const hasContinueStatement = (block: es.BlockStatement): boolean => {
  let hasContinue = false
  for (const statement of block.body) {
    if (statement.type === 'ContinueStatement') {
      hasContinue = true
    } else if (statement.type === 'IfStatement') {
      // Parser enforces that if/else have braces (block statement)
      hasContinue = hasContinue || hasContinueStatement(statement.consequent as es.BlockStatement)
      if (statement.alternate) {
        hasContinue = hasContinue || hasContinueStatement(statement.alternate as es.BlockStatement)
      }
    }
  }
  return hasContinue
}
