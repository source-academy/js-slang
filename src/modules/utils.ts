/**
 * For tabs that are export default declarations, we need to remove
 * the `export default` bit from the front before they can be loaded
 * by js-slang
 */
export function evalRawTab(text: string) {
  if (text.startsWith('export default')) {
    text = text.substring(14)
  }

  return eval(text)
}
