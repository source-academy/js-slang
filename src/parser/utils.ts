import {
  Comment,
  ecmaVersion,
  Node,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { Program, SourceLocation } from 'estree'

import { Context } from '..'
import { DEFAULT_ECMA_VERSION } from '../constants'
import { SourceError } from '../types'
import { validateAndAnnotate } from '../validator/validator'
import { MissingSemicolonError, TrailingCommaError } from './errors'
import { AcornOptions, BabelOptions } from './types'

/**
 * Generates options object for acorn parser
 *
 * @param ecmaVersion ECMA version
 * @param errors error container
 * @param throwOnError throw on error if true else push to error container and resume exec
 * @param options partial acorn options
 * @returns
 */
export const createAcornParserOptions = (
  ecmaVersion: ecmaVersion,
  errors?: SourceError[],
  options?: Partial<AcornOptions>,
  throwOnError?: boolean
): AcornOptions => ({
  ecmaVersion,
  sourceType: 'module',
  locations: true,
  onInsertedSemicolon(_tokenEndPos: number, tokenPos: Position) {
    const error = new MissingSemicolonError(positionToSourceLocation(tokenPos, options?.sourceFile))
    if (throwOnError) throw error
    errors?.push(error)
  },
  onTrailingComma(_tokenEndPos: number, tokenPos: Position) {
    const error = new TrailingCommaError(positionToSourceLocation(tokenPos, options?.sourceFile))
    if (throwOnError) throw error
    errors?.push(error)
  },
  ...options
})

/**
 * Parses a single expression at a specified offset
 *
 * @param programStr program string
 * @param offset position offset
 * @param ecmaVersion ECMA version
 * @returns acorn AST Node if parse succeeds else null
 */
export function parseAt(
  programStr: string,
  offset: number,
  ecmaVersion: ecmaVersion = DEFAULT_ECMA_VERSION
): Node | null {
  try {
    return acornParseAt(programStr, offset, { ecmaVersion })
  } catch (_error) {
    return null
  }
}

/**
 * Parse a program, returning alongside comments found within that program
 *
 * @param programStr program string
 * @param ecmaVersion ECMA version
 * @returns tuple consisting of the parsed program, and a list of comments found within the program string
 */
export function parseWithComments(
  programStr: string,
  ecmaVersion: ecmaVersion = DEFAULT_ECMA_VERSION
): [Program, Comment[]] {
  let comments: acorn.Comment[] = []
  const acornOptions: AcornOptions = createAcornParserOptions(
    ecmaVersion,
    undefined,
    {
      onComment: comments
    },
    undefined
  )

  let ast: Program | undefined
  try {
    ast = acornParse(programStr, acornOptions) as unknown as Program
  } catch {
    comments = []
    ast = acornLooseParse(programStr, acornOptions)
  }

  return [ast, comments]
}

/**
 * Parse program with error-tolerant acorn parser
 *
 * @param programStr program string
 * @param context js-slang context
 * @returns ast for program string
 */
export function looseParse(programStr: string, context: Context): Program {
  return acornLooseParse(
    programStr,
    createAcornParserOptions(DEFAULT_ECMA_VERSION, context.errors)
  ) as unknown as Program
}

/**
 * TODO
 *
 * @param programStr program string
 * @param context js-slang context
 * @returns ast for program string
 */
export function typedParse(programStr: string, context: Context): Program {
  const ast: Program = looseParse(programStr, context)
  return validateAndAnnotate(ast, context)
}

/**
 * Converts acorn parser Position object to SourceLocation object
 *
 * @param position acorn Position object
 * @returns SourceLocation
 */
export const positionToSourceLocation = (position: Position, source?: string): SourceLocation => ({
  start: { ...position },
  end: { ...position, column: position.column + 1 },
  source
})

export const defaultBabelOptions: BabelOptions = {
  sourceType: 'module',
  plugins: ['typescript', 'estree']
}
