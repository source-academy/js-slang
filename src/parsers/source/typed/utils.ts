import * as es from 'estree'

/**
 * Transforms an AST generated by the Babel parser into one
 * that is compliant with the ESTree types. While the Babel
 * parser-generated AST is mostly compliant with the ESTree
 * specifications due to the use of the 'estree' plugin, not
 * everything is compliant with the ESTree types.
 *
 * @param program The AST to be transformed.
 */
export const transformBabelASTToESTreeCompliantAST = (program: es.Program): void => {
  renameFilenameAttributeToSource(program)
}

/**
 * Renames the 'filename' attribute on AST nodes to 'source'.
 * This is because the Babel parser tags AST nodes with the
 * 'filename' attribute when the name of the file in which
 * the AST was generated from is given. However, by the ESTree
 * types that js-slang uses, this attribute should be named
 * 'source' instead of 'filename'. As a workaround, we
 * recursively walk the AST generated by the Babel parser and
 * replace all instances of the 'filename' attribute with
 * 'source'.
 *
 * Note that we do not use acorn-walk here as a precondition
 * of acorn-walk is that the AST is compliant with the ESTree
 * specifications. Since the reason why we are transforming
 * the AST generated by the Babel parser is because it is not
 * entirely compliant with the ESTree specifications in the
 * first place, we avoid the use of acorn-walk.
 *
 * @param node The node to be transformed.
 */
const renameFilenameAttributeToSource = (node: Record<string, any>): void => {
  // Rename all 'filename' attributes to 'source'.
  if (node.hasOwnProperty('filename')) {
    node.source = node.filename
    delete node.filename
  }
  // Iterate over all property values of the AST node.
  Object.values(node).forEach(value => {
    // If the value is null or undefined, do nothing.
    if (value === null || value === undefined) {
      return
    }
    // If the value is an array, check if any of its elements
    // are objects (i.e., AST nodes) which we should recurse on.
    if (Array.isArray(value)) {
      value.forEach(element => {
        if (element === null || element === undefined) {
          return
        }
        if (typeof element !== 'object') {
          return
        }
        renameFilenameAttributeToSource(element)
      })
    }
    // If the value is a primitive type, do nothing.
    if (typeof value !== 'object') {
      return
    }
    // Otherwise, the value is an object (i.e., AST node) which
    // we recurse on.
    renameFilenameAttributeToSource(value)
  })
}
