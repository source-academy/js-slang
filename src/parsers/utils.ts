import { Position, SourceLocation } from 'acorn'

export const positionToSourceLocation = (position: Position): SourceLocation => ({
  start: position,
  end: { ...position, column: position.column + 1 }
})
