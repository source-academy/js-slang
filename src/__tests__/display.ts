import { expectDisplayResult } from '../utils/testing'

test('display can be used to display numbers', () => {
  return expectDisplayResult(`display(0);`).toMatchInlineSnapshot(`
Array [
  "0",
]
`)
})

test('display can be used to display funny numbers', () => {
  return expectDisplayResult(`display(1e38); display(NaN); display(Infinity);`)
    .toMatchInlineSnapshot(`
Array [
  "1e+38",
  "NaN",
  "Infinity",
]
`)
})

test('display can be used to display (escaped) strings', () => {
  return expectDisplayResult(`display("Tom's assisstant said: \\"tuna.\\"");`)
    .toMatchInlineSnapshot(`
Array [
  "\\"Tom's assisstant said: \\\\\\"tuna.\\\\\\"\\"",
]
`)
})

test('raw_display can be used to display (unescaped) strings directly', () => {
  return expectDisplayResult(`raw_display("Tom's assisstant said: \\"tuna.\\"");`)
    .toMatchInlineSnapshot(`
Array [
  "Tom's assisstant said: \\"tuna.\\"",
]
`)
})

test('display can be used to display functions', () => {
  return expectDisplayResult(`display(x => x); display((x, y) => x + y);`).toMatchInlineSnapshot(`
Array [
  "x => x",
  "(x, y) => x + y",
]
`)
})

test('display can be used to display lists', () => {
  return expectDisplayResult(`display(list(1, 2));`, 2).toMatchInlineSnapshot(`
Array [
  "[1, [2, null]]",
]
`)
})

test('display can be used to display arrays', () => {
  return expectDisplayResult(`display([1, 2, [4, 5]]);`, 3).toMatchInlineSnapshot(`
Array [
  "[1, 2, [4, 5]]",
]
`)
})

test('display can be used to display objects', () => {
  return expectDisplayResult(`display({a: 1, b: 2, c: {d: 3}});`, 100).toMatchInlineSnapshot(`
Array [
  "{\\"a\\": 1, \\"b\\": 2, \\"c\\": {\\"d\\": 3}}",
]
`)
})
