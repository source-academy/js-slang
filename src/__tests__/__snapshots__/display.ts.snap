// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display can be used to display (escaped) strings: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display(\\"Tom's assisstant said: \\\\\\"tuna.\\\\\\"\\");",
  "displayResult": Array [
    "\\"Tom's assisstant said: \\\\\\"tuna.\\\\\\"\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": "Tom's assisstant said: \\"tuna.\\"",
  "resultStatus": "finished",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
const unaryOp = native.operators.get(\\"evaluateUnaryExpressionIfValidElseError\\");
const binaryOp = native.operators.get(\\"evaluateBinaryExpressionIfValidElseError\\");
const throwIfTimeout = native.operators.get(\\"throwIfExceedsTimeLimit\\");
(( <globals redacted> ) => {
  return (() => {
    const lastStatementResult = eval(\\"callIfFuncAndRightArgs(display, 1, 0, \\\\\\"Tom's assisstant said: \\\\\\\\\\\\\\"tuna.\\\\\\\\\\\\\\"\\\\\\");\\");
    return lastStatementResult;
  })();
})();
",
  "visualiseListResult": Array [],
}
`;

exports[`display can be used to display arrays: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display([1, 2, [4, 5]]);",
  "displayResult": Array [
    "[1, 2, [4, 5]]",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": Array [
    1,
    2,
    Array [
      4,
      5,
    ],
  ],
  "resultStatus": "finished",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
const unaryOp = native.operators.get(\\"evaluateUnaryExpressionIfValidElseError\\");
const binaryOp = native.operators.get(\\"evaluateBinaryExpressionIfValidElseError\\");
const throwIfTimeout = native.operators.get(\\"throwIfExceedsTimeLimit\\");
(( <globals redacted> ) => {
  return (() => {
    const lastStatementResult = eval(\\"callIfFuncAndRightArgs(display, 1, 0, [1, 2, [4, 5]]);\\");
    return lastStatementResult;
  })();
})();
",
  "visualiseListResult": Array [],
}
`;

exports[`display can be used to display functions: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display(x => x); display((x, y) => x + y);",
  "displayResult": Array [
    "x => x",
    "(x, y) => x + y",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": [Function],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`display can be used to display funny numbers: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display(1e38); display(NaN); display(Infinity);",
  "displayResult": Array [
    "1e+38",
    "NaN",
    "Infinity",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": Infinity,
  "resultStatus": "finished",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
const unaryOp = native.operators.get(\\"evaluateUnaryExpressionIfValidElseError\\");
const binaryOp = native.operators.get(\\"evaluateBinaryExpressionIfValidElseError\\");
const throwIfTimeout = native.operators.get(\\"throwIfExceedsTimeLimit\\");
(( <globals redacted> ) => {
  return (() => {
    callIfFuncAndRightArgs(display, 1, 0, 1e38);
    callIfFuncAndRightArgs(display, 1, 15, NaN);
    const lastStatementResult = eval(\\"callIfFuncAndRightArgs(display, 1, 29, Infinity);\\");
    return lastStatementResult;
  })();
})();
",
  "visualiseListResult": Array [],
}
`;

exports[`display can be used to display lists: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display(list(1, 2));",
  "displayResult": Array [
    "[1, [2, null]]",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": Array [
    1,
    Array [
      2,
      null,
    ],
  ],
  "resultStatus": "finished",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
const unaryOp = native.operators.get(\\"evaluateUnaryExpressionIfValidElseError\\");
const binaryOp = native.operators.get(\\"evaluateBinaryExpressionIfValidElseError\\");
const throwIfTimeout = native.operators.get(\\"throwIfExceedsTimeLimit\\");
(( <globals redacted> ) => {
  return (() => {
    const lastStatementResult = eval(\\"callIfFuncAndRightArgs(display, 1, 0, callIfFuncAndRightArgs(list, 1, 8, 1, 2));\\");
    return lastStatementResult;
  })();
})();
",
  "visualiseListResult": Array [],
}
`;

exports[`display can be used to display numbers: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display(0);",
  "displayResult": Array [
    "0",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 0,
  "resultStatus": "finished",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
const unaryOp = native.operators.get(\\"evaluateUnaryExpressionIfValidElseError\\");
const binaryOp = native.operators.get(\\"evaluateBinaryExpressionIfValidElseError\\");
const throwIfTimeout = native.operators.get(\\"throwIfExceedsTimeLimit\\");
(( <globals redacted> ) => {
  return (() => {
    const lastStatementResult = eval(\\"callIfFuncAndRightArgs(display, 1, 0, 0);\\");
    return lastStatementResult;
  })();
})();
",
  "visualiseListResult": Array [],
}
`;

exports[`display can be used to display objects: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display({a: 1, b: 2, c: {d: 3}});",
  "displayResult": Array [
    "{\\"a\\": 1, \\"b\\": 2, \\"c\\": {\\"d\\": 3}}",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": Object {
    "a": 1,
    "b": 2,
    "c": Object {
      "d": 3,
    },
  },
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`raw_display can be used to display (unescaped) strings directly: expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "raw_display(\\"Tom's assisstant said: \\\\\\"tuna.\\\\\\"\\");",
  "displayResult": Array [
    "Tom's assisstant said: \\"tuna.\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": "Tom's assisstant said: \\"tuna.\\"",
  "resultStatus": "finished",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
const unaryOp = native.operators.get(\\"evaluateUnaryExpressionIfValidElseError\\");
const binaryOp = native.operators.get(\\"evaluateBinaryExpressionIfValidElseError\\");
const throwIfTimeout = native.operators.get(\\"throwIfExceedsTimeLimit\\");
(( <globals redacted> ) => {
  return (() => {
    const lastStatementResult = eval(\\"callIfFuncAndRightArgs(raw_display, 1, 0, \\\\\\"Tom's assisstant said: \\\\\\\\\\\\\\"tuna.\\\\\\\\\\\\\\"\\\\\\");\\");
    return lastStatementResult;
  })();
})();
",
  "visualiseListResult": Array [],
}
`;
